File: Laboratories/Lab 2 - C++/Buffer.cpp
//
// Created by night on 11/1/2022.
//

#include "Buffer.h"
#include <iostream>
#include <thread>
Buffer::Buffer() {

}

int Buffer::get() {

    std::unique_lock<std::mutex> lock(mutex);

    while(queue.empty()){
        std::cout<<"Thread with id " << std::this_thread::get_id()<<" : Buffer is currently empty"<<"\n";
        readyToReceiveProduct.wait(lock);
    }

    int value = queue.front();
    queue.pop();

    std::cout<<"Thread with id " <<std::this_thread::get_id()<<" extracted value " << value << " from the queue\n";
    readyToSendProduct.notify_one();

    return value;
}

void Buffer::put(int val) {

    std::unique_lock<std::mutex> lock(mutex);

    while(queue.size() == 1) {
        std::cout<<"Thread with id " << std::this_thread::get_id() << " : Queue is currently full" << "\n";
        readyToSendProduct.wait(lock);
    }

    queue.push(val);

    std::cout<<"Thread with id " << std::this_thread::get_id()<<" added value " << val << " to the queue\n";
    readyToReceiveProduct.notify_one();

}

File: Laboratories/Lab 2 - Java/src/Buffer.java
import java.util.LinkedList;
import java.util.Queue;import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import static java.sql.Types.NULL;

public class Buffer {

    private final Queue<Integer> queue = new LinkedList<>();
    private final Lock lock = new ReentrantLock();

    private final Condition readyToSendProduct = lock.newCondition();

    private final Condition readyToReceiveProduct = lock.newCondition();

    private Integer partialResult = NULL;

    public Buffer(){

    }
    public int get() throws InterruptedException{

        lock.lock();

        try{
            while(partialResult == NULL){
                System.out.println(Thread.currentThread().getName() + ": Buffer is currently empty");
                readyToReceiveProduct.await();
            }

            Integer value = queue.poll();
            if(value != null){
                partialResult = NULL;
                System.out.printf("%s extracted value %d from the queue\n", Thread.currentThread().getName(), value);
                readyToSendProduct.signal();
            }
            return value;
        } finally {
            lock.unlock();
        }
    }

    public void put(int val) throws InterruptedException{

        lock.lock();

        try{
            while (partialResult != NULL){
                System.out.println(Thread.currentThread().getName() + ": Queue is currently full");
                readyToSendProduct.await();
            }

            queue.add(val);
            partialResult = val;
            System.out.printf("%s added value %d to the queue\n", Thread.currentThread().getName(), val);
            readyToReceiveProduct.signal();

        }
        finally {
            lock.unlock();
        }
    }



}
